plugins/inputs/azure_monitor/azure_monitor.go

```diff
@@ -0,0 +1,856 @@
+	request.Header.Set("Authorization", "Bearer "+am.azureClient.accessToken)
+
+	response, err := am.azureClient.client.Do(request)
+
+	if err != nil {
+		return nil, fmt.Errorf("error getting response from Azure Monitor API: %v", err)
+	}
+
+	defer closeResponseBody(response.Body, &err)
+
+	body, err := getResponseBody(response)
+
+	return body, err
+}
+
+func (am *AzureMonitor) collectResourceTargetsMetrics(acc telegraf.Accumulator) error {
+	var waitGroup sync.WaitGroup
+
+	errChan := make(chan error, 1)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	defer cancel()
+
+	for index, target := range am.ResourceTargets {
+		am.Log.Info("Collecting metrics for target #", index+1)
+		waitGroup.Add(1)
+
+		go func(target *ResourceTarget) {
+			defer waitGroup.Done()
+
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			apiURL := am.buildMetricValuesAPIURL(target)
+			body, err := am.getTargetResponseBody(apiURL)
+
+			if err != nil {
+				select {
+				case errChan <- err:
+				default:
+				}
+
+				cancel()
+				return
+			}
+
+			metrics, err := am.collectResourceTargetMetrics(body)
+
+			if err != nil {
+				select {
+				case errChan <- err:
+				default:
+				}
+
+				cancel()
+				return
+			}
+
+			for _, metric := range metrics {
+				acc.AddFields(metric.name, metric.fields, metric.tags, time.Now())
+			}
+		}(target)
+
+		waitGroup.Wait()
+
+		select {
+		case err := <-errChan:
+			return err
+		default:
+		}
+	}
+
+	return nil
+}
+
+func (am *AzureMonitor) collectResourceTargetMetrics(body []byte) ([]*Metric, error) {
+	bodyData, err := unmarshalJSON(body)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var metrics []*Metric
+
+	for _, value := range bodyData["value"].([]interface{}) {
+		var metric *Metric
+
+		timeSeries := value.(map[string]interface{})["timeseries"].([]interface{})
+
+		if len(timeSeries) == 0 {
+			metricName, fullResourceName, resourceType := getMetricDetails(value.(map[string]interface{}))
+
+			am.Log.Info("There is no timeSeries to metric: ", metricName, " for resource: ", fullResourceName, " type: ", resourceType)
+			continue
+		}
+
+		data := timeSeries[0].(map[string]interface{})["data"].([]interface{})
+
+		if len(data) == 0 {
+			metricName, fullResourceName, resourceType := getMetricDetails(value.(map[string]interface{}))
+
+			am.Log.Info("There is no value to metric: ", metricName, " for resource: ", fullResourceName, " type: ", resourceType)
+			continue
+		}
+
+		if !isMetricHaveValue(data) {
+			metricName, fullResourceName, resourceType := getMetricDetails(value.(map[string]interface{}))
+
+			am.Log.Info("There is no value to metric: ", metricName, " for resource: ", fullResourceName, " type: ", resourceType)
+			continue
+		}
+
+		metric = NewMetric()
+
+		metric.getMetricName(value.(map[string]interface{}))
+		metric.getMetricFields(data)
+		metric.getMetricTags(bodyData, value.(map[string]interface{}))
+
+		metrics = append(metrics, metric)
```

Comment:
Can we please directly collect `telegraf.Metric`s here instead of the custom structure definition?